### PR TITLE
Formtoolkit

### PR DESCRIPTION
**Pull request by chapoi**
This PR introduces FormKit, a component-based form library designed to simplify form creation and management. This library provides a single `Form` component, various field components, controls, validation mechanisms, and customization options. Additionally, it includes helpers to facilitate testing and writing specifications for forms.

1. **Form Component**:
   - The main component that encapsulates form logic and structure.
   - Yields various utilities like `Field`, `Submit`, `Alert`, etc.

   **Example Usage**:
   ```gjs
   import Form from "discourse/form";

   <template>
     <Form as |form|>
       <form.Field
         @name="username"
         @title="Username"
         @validation="required"
         as |field|
       >
         <field.Input />
       </form.Field>

       <form.Field @name="age" @title="Age" as |field|>
         <field.Input @type="number" />
       </form.Field>

       <form.Submit />
     </Form>
   </template>
   ```

2. **Validation**:
   - Built-in validation rules such as `required`, `number`, `length`, and `url`.
   - Custom validation callbacks for more complex validation logic.

   **Example Usage**:
   ```javascript
   validateUsername(name, value, data, { addError }) {
     if (data.bar / 2 === value) {
       addError(name, "That's not how maths work.");
     }
   }
   ```

   ```hbs
   <form.Field @name="username" @validate={{this.validateUsername}} />
   ```

3. **Customization**:
   - Plugin outlets for extending form functionality.
   - Styling capabilities through propagated attributes.
   - Custom controls with properties provided by `form` and `field`.

   **Example Usage**:
   ```hbs
   <Form class="my-form" as |form|>
     <form.Field class="my-field" as |field|>
       <MyCustomControl id={{field.id}} @onChange={{field.set}} />
     </form.Field>
   </Form>
   ```

4. **Helpers for Testing**:
   - Test assertions for form and field validation.

   **Example usage**:
   ```javascript
   assert.form().hasErrors("the form shows errors");
   assert.form().field("foo").hasValue("bar", "user has set the value");
   ```

   - Helper for interacting with he form

   **Example usage**:
   ```javascript
   await formKit().field("foo").fillIn("bar");
   ```

5. **Page Object for System Specs**:
   - Page objects for interacting with forms in system specs.
   - Methods for submitting forms, checking alerts, and interacting with fields.

   **Example Usage**:
   ```ruby
   form = PageObjects::Components::FormKit.new(".my-form")
   form.submit
   expect(form).to have_an_alert("message")
   ```

   **Field Interactions**:
   ```ruby
   field = form.field("foo")
   expect(field).to have_value("bar")
   field.fill_in("bar")
   ```

6. **Collections handling**:
   - A specific component to handle array of objects

   **Example Usage**:
   ```gjs
    <Form @data={{hash foo=(array (hash bar=1) (hash bar=2))}} as |form|>
      <form.Collection @name="foo" as |collection|>
        <collection.Field @name="bar" @title="Bar" as |field|>
          <field.Input />
        </collection.Field>
      </form.Collection>
    </Form>
   ```

### CODE REVIEW

File: app/assets/javascripts/admin/addon/components/modal/badge-preview.js
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/modal/badge-preview.js:45]: This was already set on parent

File: app/assets/javascripts/admin/addon/controllers/admin-badges/show.js
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/controllers/admin-badges/show.js:1]: the biggest change here is to have stopped using `bufferedProperty` given form-kit includes its own proxy

File: app/assets/javascripts/form-kit/addon/components/control-wrapper.gjs
Comment (jjaffeux) [inline app/assets/javascripts/form-kit/addon/components/control-wrapper.gjs:65]: these properties are a problem because they are only useful to a specific component and not all of them

File: app/assets/javascripts/form-kit/addon/components/control/input.gjs
Comment (davidtaylorhq) [inline app/assets/javascripts/form-kit/addon/components/control/input.gjs:38]: BTW, these Ember assertions are optimised out of production builds. So if people are developing themes against production sites, they will never be hit. Maybe fine, but just want to mention.

Comment (jjaffeux) [reply app/assets/javascripts/form-kit/addon/components/control/input.gjs:38]: Yes I know how it works but indeed I didnt think of people developing themes directly against a production build. That was actually a question on my list to ask you why we don't use assert more, I guess I know now :p

File: app/assets/javascripts/discourse/app/form-kit/components/fk/form.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/form-kit/components/fk/form.gjs:177]: I would like to avoid this, but couldn't figure out how to without having to have many `settled()` in my tests

File: app/assets/javascripts/discourse/app/form-kit/components/fk/form.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/form-kit/components/fk/form.gjs:231]: I would like to avoid this, but couldn't figure out how to without having to have many `settled()` in my tests

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:147]: I thought I had removed it, you should never write any formkit css, it's an admin-badge thing here, I will change it thanks :+1:

File: app/assets/javascripts/discourse/app/form-kit/components/fk/char-counter.gjs
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/char-counter.gjs:14]: ah yes good idea :+1: I need to write tests for this too, haven't done it yet

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:86]: Will think about it :+1:

File: app/assets/javascripts/admin/addon/controllers/admin-badges/show.js
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/controllers/admin-badges/show.js:111]: Not sure I can do this, I already did it part of it on the Submit, but not sure I can go further, will think about it.

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:87]: I fixed the other comments, but will wait for after merge on this one, I didnt document this specific component, so I consider it private API for now, will try to improve it after.

**Comment by jjaffeux** [comment]
The failing spec in discourse-ai is due to me moving ace-editor component out of admin as Im using it for the code component.

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:1]: The first usage of form kit on a core form

File: app/assets/javascripts/discourse/app/components/ace-editor.js
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/ace-editor.js:131]: I needed to have an onChange event for the code component, everything in FormKit is following DDAU, there's no implicit mutation

File: app/assets/javascripts/discourse/app/components/ace-editor.js
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/ace-editor.js:150]: This is fixing a mistake in core currently where we set a whole object on a data attribute

File: app/assets/javascripts/discourse/app/components/d-editor.hbs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-editor.hbs:70]: For accessibility to work correctly it's important to have ids

File: app/assets/javascripts/discourse/app/components/d-toggle-switch.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-toggle-switch.gjs:26]: form kit provides its own label

File: app/assets/javascripts/discourse/app/components/form.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/form.gjs:1]: A proxy to the namespaced library in discourse/form-kit

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs:20]: For now mostly input type=text is supported, I will add more as I convert more forms

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs:88]: This allow errors to create anchors to specific fields

File: app/assets/javascripts/admin/addon/controllers/admin-badges/show.js
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/controllers/admin-badges/show.js:17]: FormKit now uses its own change set (backed by immer.js) have the same behavior than `bufferedProperty` was offering. On top of this it also support nested paths.

If you are not familiar with `bufferedProperty`, this is a mechanism which allows to store mutations on a proxy object without mutating the original object. You can decide to apply these mutations later.

File: app/assets/javascripts/admin/addon/controllers/admin-badges/show.js
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/controllers/admin-badges/show.js:46]: FormKit only works with POJO (this is actually a limitation of immer.js but I think there's value into this anyways).

File: app/assets/javascripts/admin/addon/routes/admin-badges/show.js
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/routes/admin-badges/show.js:56]: This has been moved into the controller, to avoid the need for `badge.set("preview_loading", ...)` which is not possible with FormKit as we don't manipulate the model object and also because it's not a good pattern.

File: app/assets/javascripts/admin/addon/templates/admin-badges.hbs
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/templates/admin-badges.hbs:2]: This is UX changes that @chapoi wanted

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control/radio-group.gjs
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/form-kit/components/fk/control/radio-group.gjs:6]: I can use template only components, as I needed a class name for type checking

File: app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs:71]: No, it's only if you do `<form.FIeld @size={{4}} />`, it's a shortcut

File: app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs:64]: fields can include dots, this is how I manage collections `COLLECTION_NAME.INDEX.FIELD_NAME`

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs:14]: I cant do exactly this because input is dynamic with type, but I made a mix of this :+1:

File: app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs:64]: Then I think we should make our own helper for digging into nested objects based on a string. EmberObject's get has a ton of extra complexity which may affect autotracking & performance.

File: app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs:75]: Oh right, I see 👌

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control-wrapper.gjs:36]: Names now cannot include `-` 👌

File: plugins/styleguide/assets/javascripts/discourse/components/sections/atoms/05-input-fields.hbs
Comment (davidtaylorhq) [inline plugins/styleguide/assets/javascripts/discourse/components/sections/atoms/05-input-fields.hbs:1]: What's the plan here? I guess we should either add something, or remove it entirely?

File: app/assets/javascripts/discourse/app/form-kit/lib/fk-form-data.js
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/form-kit/lib/fk-form-data.js:182]: Would be good to replace this Ember `set`, like we did for `get`

File: plugins/styleguide/assets/javascripts/discourse/components/sections/atoms/05-input-fields.hbs
Comment (jjaffeux) [reply plugins/styleguide/assets/javascripts/discourse/components/sections/atoms/05-input-fields.hbs:1]: I need to find the motivation to write it :p

File: app/assets/javascripts/discourse/app/form-kit/lib/fk-form-data.js
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/lib/fk-form-data.js:188]: ah yes good point

Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/lib/fk-form-data.js]: This is now done

File: app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/field.gjs]: this is now done

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control/radio-group.gjs
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control/radio-group.gjs:6]: I removed the comment. eslint is happy now that the class has a static property

File: app/assets/javascripts/admin/addon/routes/admin-badges.js
Comment (davidtaylorhq) [inline app/assets/javascripts/admin/addon/routes/admin-badges.js:23]:
```suggestion
    const model = this.controllerFor("admin-badges").badgeGroupings;
```

**Review by davidtaylorhq — APPROVED** [review]
We talked about some issues with validation triggering too early. But otherwise, this looks good to me ✅

**Comment by jjaffeux** [comment]
> We talked about some issues with validation triggering too early. But otherwise, this looks good to me ✅

this is fixed

**Pull request merged**
